nemoguardrails/embeddings/cache.py

```diff
@@ -0,0 +1,327 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import functools
+import hashlib
+import json
+import logging
+from abc import ABC, abstractmethod
+from functools import singledispatchmethod
+from pathlib import Path
+from typing import List
+
+from nemoguardrails.rails.llm.config import EmbeddingsCacheConfig
+
+log = logging.getLogger(__name__)
+
+
+class KeyGenerator(ABC):
```

Comment:
It is possible to enforce name class variable using Meta
```python
class KeyGeneratorMeta(ABCMeta):
    """Metaclass for KeyGenerator."""

    def __init__(cls, name, bases, attrs):
        super().__init__(name, bases, attrs)
        if not inspect.isabstract(cls):  # Check if it's a concrete class
            if not hasattr(cls, 'name'):
                raise TypeError("Subclasses of KeyGenerator must define a 'name' class variable")
            if 'name' in attrs:
                # Check that the name is unique among all subclasses of KeyGenerator
                for other in KeyGenerator.__subclasses__():
                    if other is not cls and other.name == attrs['name']:
                        raise ValueError(f"Duplicate KeyGenerator name: {attrs['name']}")


class KeyGenerator(ABC, metaclass=KeyGeneratorMeta):
    """Abstract class for key generators."""
  
  ```
    
    or generate the name based on class name using following utils:
    
    ```python
    import re 

def to_snake_case(name):
    name = re.sub(r"\W+", "", name)
    name = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", name)
    name = re.sub("([a-z])([A-Z])", r"\1_\2", name).lower()
    return name



def get_object_name(obj):
    if hasattr(obj, "name"):
        return obj.name
    elif hasattr(obj, "__name__"):
        return to_snake_case(obj.__name__)
    elif hasattr(obj, "__class__"):
        return to_snake_case(obj.__class__.__name__)
    return to_snake_case(str(obj))
    ```
    as below
    
    ```python
        @classmethod
    def from_name(cls, name):
        for subclass in cls.__subclasses__():
            # Use the 'name' attribute if it's defined, otherwise use the class name
            subclass_name = get_object_name(subclass).replace(cls.__name__, "")
            if subclass_name == name:
                return subclass
        raise ValueError(f"Unknown key generator: {name}")
    ```